lib/screens/v2/candidate_generator/dup.ex

```diff
@@ -113,86 +134,212 @@ defmodule Screens.V2.CandidateGenerator.Dup do
-        ]
-      end
+      sections_data_to_departure_instances(
+        config,
+        primary_sections_data,
+        [:main_content_zero, :main_content_one],
+        now
+      )
 
     secondary_departures_instances =
-      if Enum.any?(secondary_sections_data, &(&1 == :error)) do
-        %DeparturesNoData{screen: config, show_alternatives?: true}
-      else
-        sections =
-          Enum.map(secondary_sections_data, fn {:ok, departures} ->
-            visible_departures =
-              if length(secondary_sections_data) > 1 do
-                Enum.take(departures, 2)
-              else
-                Enum.take(departures, 4)
-              end
-
-            %{type: :normal_section, rows: visible_departures}
-          end)
-
-        [
-          %DeparturesWidget{
-            screen: config,
-            section_data: sections,
-            slot_names: [:main_content_two]
-          }
-        ]
-      end
+      sections_data_to_departure_instances(
+        config,
+        secondary_sections_data,
+        [:main_content_two],
+        now
+      )
 
     primary_departures_instances ++ secondary_departures_instances
   end
 
+  defp sections_data_to_departure_instances(config, sections_data, slot_ids, now) do
+    if Enum.any?(sections_data, &(&1 == :error)) do
+      %DeparturesNoData{screen: config, show_alternatives?: true}
+    else
+      sections =
+        Enum.map(sections_data, fn %{
+                                     departures: departures,
+                                     pill: pill,
+                                     alert: alert,
+                                     headway: headway,
+                                     stop_ids: stop_ids
+                                   } ->
+          visible_departures =
+            if length(sections_data) > 1 do
+              Enum.take(departures, 2)
+            else
+              Enum.take(departures, 4)
+            end
+
+          case fetch_headway_mode(stop_ids, headway, alert, now) do
+            {:active, time_range, headsign} ->
+              %{type: :headway_section, pill: pill, time_range: time_range, headsign: headsign}
+
+            :inactive ->
+              %{type: :normal_section, rows: visible_departures}
+          end
+        end)
+
+      Enum.map(slot_ids, fn slot_id ->
+        %DeparturesWidget{
+          screen: config,
+          section_data: sections,
+          slot_names: [slot_id]
+        }
+      end)
+    end
+  end
+
+  defp get_sections_data(sections, fetch_section_departures_fn, fetch_alerts_or_empty_list_fn) do
+    Enum.map(sections, fn %Section{
+                            query: %Query{params: %Params{stop_ids: stop_ids} = params},
+                            headway: %Headway{pill: pill} = headway
+                          } = section ->
+      section_alert = get_section_alert(params, fetch_alerts_or_empty_list_fn)
+
+      {:ok, section_departures} = section |> fetch_section_departures_fn.()
```

Comment:
If the departures fetch chokes and returns an error, this pattern match fails. (See that Widget.Departures.filter_departures supports returning an error.)